### PR TITLE
Partial solutions

### PR DESCRIPTION
**Pull request by rrsadykov**
- Partial solutions are introduced
- Added support of partial solutions in ColumnGeneration, SolveIpForm, SolveLpForm algorithms
- Correction in calculation of column reduced costs
- Corrections in preprocessing
- Correction in storages restoration
- New test with a user defined algorithm (to test partial solutions and preprocessing)

Branch name is an old one, no diving here.

### CODE REVIEW

**Comment by rrsadykov** [comment]
@guimarqu Can you please see the preprocessing tests on Julia 1.0. It says `type CachingOptimizer has no field inner`
for the lines
```
coluna_optimizer = problem.moi_backend
master = CL.getmaster(coluna_optimizer.inner.re_formulation)

```

**Comment by rrsadykov** [comment]
The tests pass for Julia 1.5, but do not pass for Julia 1.0. The tests were done by Teobaldo I suppose, I do not understand them much

**Comment by guimarqu** [comment]
Hi Ruslan, I'll look the PR monday


> @guimarqu Can you please see the preprocessing tests on Julia 1.0. It says `type CachingOptimizer has no field inner`
> for the lines
> 
> ```
> coluna_optimizer = problem.moi_backend
> master = CL.getmaster(coluna_optimizer.inner.re_formulation)
> ```

This bug might be related to issue #376. Is it only with Julia 1.0 ?

**Comment by guimarqu** [comment]
Tests have passed with julia 1.0 : https://travis-ci.com/github/atoptima/Coluna.jl/builds/189223931

**Comment by rrsadykov** [comment]
Yes, it is strange, now tests pass, I have corrected one thing which is not supported in Julia 1.0 in the new test. 

However, the preprocessing tests still do not pass on my computer with Julia 1.0., but they pass with Julia 1.5. I agree with the issue #376 that we need to update these tests. May be it is better to remove these preprocessing tests. Now preprocessing is tested in the new test I have added (user_algorithms_tests).

**Comment by guimarqu** [comment]
Ok so let's comment them in the meantime.
‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐

**Comment by guimarqu** [comment]
Do you have the last version of BlockDecomposition in your julia 1.0 env ?

‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐

**Comment by rrsadykov** [comment]
I have updated BlockDecomposition, still the preprocessing tests do not pass on on my machine with Julia 1.0.

There are also warnings:
```
WARNING: could not import MathProg.getfeasibilitystatus into Algorithm
WARNING: could not import MathProg.setfeasibilitystatus! into Algorithm
WARNING: could not import MathProg.isfeasible into Algorithm
```  
They are related to your previous merge

**Comment by rrsadykov** [comment]
I have started to look at the preprocessing in more detail. There are a lot of corrections to make. I will submit them in separate pull request (the old preprocessing tests will be deactivated there). I think we can merge this pull request as it is now, as it passes all checks.

**Review by guimarqu — COMMENTED** [review]
I won't merge this branch into master. I'll create a protected branch for v0.4 in which we will merge partial solutions and non-robust cuts so we can gradually improve the code without breaking things in 0.3.

File: src/Algorithm/Algorithm.jl [hunk 1]
```diff
@@ -16,8 +16,8 @@ const MOI = MathOptInterface
 import Base: push!
 
 # Import to extend methods to OptimizationState
-import ..MathProg: getterminationstatus,
-    setterminationstatus!, get_ip_primal_bound, get_ip_dual_bound,
+import ..MathProg: getfeasibilitystatus, getterminationstatus, setfeasibilitystatus!,
+    setterminationstatus!, isfeasible, get_ip_primal_bound, get_ip_dual_bound,
```
[the remark this answers is not in the document]
Comment (guimarqu) [reply src/Algorithm/Algorithm.jl:20]: I think it's related to your comment about the WARNING. These methods have been deleted.

[the remark this answers is not in the document]
Comment (rrsadykov) [reply src/Algorithm/Algorithm.jl:20]: I do not remember I added this, but never mind, I will delete.

File: test/Project.toml [hunk 2]
```diff
@@ -8,3 +8,4 @@ MathOptInterface = "b8f27783-ece8-5eb3-8dc8-9495eed66fee"
 Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
 Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
 LightGraphs = "093fc24a-ae57-5d10-9952-331d41423f4d"
+Parameters = "d96e819e-fc66-5662-9728-84c9c7592b0a"
```
[the remark this answers is not in the document]
Comment (rrsadykov) [reply test/Project.toml:11]: This to make a cleaner code for the user algorithm test (to use `@with_kw`).

File: src/Algorithm/colgen.jl [hunk 3]
```diff
@@ -296,7 +299,7 @@ function solve_sp_to_gencol!(
                         push!(spinfo.sol_ids_to_activate, col_id)
                     else
                         msg = """
-                        Column already exists as $(getname(masterform, col_id)) and is already active.
+                            Column already exists as $(getname(masterform, col_id)) and is already active.
```
[the remark this answers is not in the document]
Comment (rrsadykov) [reply src/Algorithm/colgen.jl:302]: ok

File: src/Algorithm/colgen.jl [hunk 4]
```diff
@@ -284,6 +286,7 @@ function solve_sp_to_gencol!(
     sense = getobjsense(masterform)
     if nb_ip_primal_sols(sp_optstate) > 0
         bestsol = get_best_ip_primal_sol(sp_optstate)
+        #@show bestsol compute_red_cost(algo, masterform, spinfo, bestsol, dualsol)
```
[the remark this answers is not in the document]
Comment (rrsadykov) [reply src/Algorithm/colgen.jl:289]: ok

File: src/Algorithm/colgen.jl [hunk 5]
```diff
@@ -254,7 +255,8 @@ function compute_red_cost(
     else
         red_cost = getvalue(spsol)
     end
-    red_cost -= (spinfo.lb * spinfo.lb_dual + spinfo.ub * spinfo.ub_dual)
+    #red_cost -= (spinfo.lb * spinfo.lb_dual + spinfo.ub * spinfo.ub_dual)
```
[the remark this answers is not in the document]
Comment (rrsadykov) [reply src/Algorithm/colgen.jl:258]: I will remove

File: src/Algorithm/colgen.jl [hunk 6]
```diff
@@ -21,7 +21,7 @@ restricted master and `pricing_prob_solve_alg` to solve the subproblems.
 
 """
 @with_kw struct ColumnGeneration <: AbstractOptimizationAlgorithm
-    restr_master_solve_alg = SolveLpForm(get_dual_solution = true)
+    restr_master_solve_alg = SolveRestrMasterInColGen()
```
[the remark this answers is not in the document]
Comment (rrsadykov) [reply src/Algorithm/colgen.jl:24]: ok

File: src/Algorithm/basic/solveipform.jl [hunk 7]
```diff
@@ -56,9 +60,17 @@ function run!(algo::SolveIpForm, data::ModelData, input::OptimizationInput)::Opt
 
     setterminationstatus!(optstate, getterminationstatus(optimizer_result))
 
-    bestprimalsol = getbestprimalsol(optimizer_result)
+    if isa(form, Formulation{MathProg.DwMaster})
+        partsolstorage = getstorage(data, PartialSolutionStoragePair)
+        partial_solution = get_primal_solution(partsolstorage, form)
+    else    
+        partial_solution = EmptyPrimalSolution(form)
```
[the remark this answers is not in the document]
Comment (rrsadykov) [reply src/Algorithm/basic/solveipform.jl:67]: Why two arrays?
I agree that most of the time, partial solution is empty. But I do not know how to make it lighter. If you have an idea..

File: src/Algorithm/basic/solveipform.jl [hunk 8]
```diff
@@ -73,11 +85,15 @@ function run!(algo::SolveIpForm, data::ModelData, input::OptimizationInput)::Opt
...
     if algo.get_dual_bound && getterminationstatus(optimizer_result) == OPTIMAL
         # TO DO : dual bound should be set in optimizer_result
-        set_ip_dual_bound!(optstate, DualBound(form, getvalue(getprimalbound(optimizer_result))))
+        dual_bound = getvalue(getprimalbound(optimizer_result)) + getvalue(partial_solution)
+        set_ip_dual_bound!(optstate, DualBound(form, dual_bound))
     end
     return OptimizationOutput(optstate)
 end
 
+run!(algo::SolveIpForm, data::ReformData, input::OptimizationInput) =
```
[the remark this answers is not in the document]
Comment (rrsadykov) [reply src/Algorithm/basic/solveipform.jl:94]: This is to call primal heuristics from the conquer algorithm. Some heuristics may need to work with reformulation. So the restricted master IP heuristic should support the call with `ReformData`.

File: src/Algorithm/conquer.jl [hunk 9]
```diff
@@ -186,7 +218,7 @@ end
 ####################################################################
 
 @with_kw struct RestrMasterLPConquer <: AbstractConquerAlgorithm 
-    masterlpalgo::SolveLpForm = SolveLpForm()
+    masterlpalgo::SolveLpForm = SolveLpFormInRestrMasterLPConquer()
```
[the remark this answers is not in the document]
Comment (rrsadykov) [reply src/Algorithm/conquer.jl:221]: ok

File: src/MathProg/solutions.jl [hunk 10]
```diff
@@ -49,6 +53,24 @@ function contains(sol::DualSolution, f::Function)
     return false
 end
 
+function concatenate_sols(sola::PrimalSolution{M}, solb::PrimalSolution{M}) where {M}
```
[the remark this answers is not in the document]
Comment (rrsadykov) [reply src/MathProg/solutions.jl:56]: yes, why not.

File: src/MathProg/formulation.jl [hunk 11]
```diff
@@ -637,3 +637,24 @@ function write_to_LP_file(form::Formulation, filename::String)
         MOI.write_to_file(dest, filename)
     end
 end
+
+function getspsol(master::Formulation{DwMaster}, col_id::VarId)
```
[the remark this answers is not in the document]
Comment (rrsadykov) [reply src/MathProg/formulation.jl:641]: I was using it while debugging. Can we just comment?

**Comment by rrsadykov** [comment]
I will do the corrections tomorrow

File: test/Project.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (guimarqu) [reply test/Project.toml:11]: Ah sorry, it's in test so it's ok !

File: src/Algorithm/basic/solveipform.jl (hunk 7 above)
[the remark this answers is not in the document]
Comment (guimarqu) [reply src/Algorithm/basic/solveipform.jl:67]: This what `EmptyPrimalSolution()` does. We can keep it like that for now. We will try to improve this gradually.

File: src/Algorithm/basic/solveipform.jl (hunk 8 above)
[the remark this answers is not in the document]
Comment (guimarqu) [reply src/Algorithm/basic/solveipform.jl:94]: Ok

File: src/MathProg/solutions.jl (hunk 10 above)
[the remark this answers is not in the document]
Comment (guimarqu) [reply src/MathProg/solutions.jl:56]: Ok we keep this for future PR.

File: src/MathProg/formulation.jl (hunk 11 above)
[the remark this answers is not in the document]
Comment (guimarqu) [reply src/MathProg/formulation.jl:641]: Ok

File: src/Algorithm/utilities/optimizationstate.jl [hunk 12]
```diff
@@ -127,7 +127,8 @@ end
 getform(state::OptimizationState{F,S}) where {F, S} = F
 
 function CopyBoundsAndStatusesFromOptState(
-    form::AbstractFormulation, source_state::OptimizationState, copy_ip_primal_sol::Bool
+    form::AbstractFormulation, source_state::OptimizationState,
```
[the remark this answers is not in the document]
Comment (rrsadykov) [reply src/Algorithm/utilities/optimizationstate.jl:130]: I have renamed this function to `OptimizationState` I have also commented previous function in this file, as it was not used

**Comment by rrsadykov** [comment]
Corrections are made

**Pull request merged**